{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import glob\n",
    "import mygene \n",
    "# os.chdir(os.path.join('tissue_cocoputs','data'))\n",
    "\n",
    "# print(os.listdir()\n",
    "# )\n",
    "# human_cds = pd.read_csv('o586358-Human_CDS_Bicod.tsv')\n",
    "\n",
    "# human_cds\n",
    "\n",
    "def subset_and_write(df,col_list,fn):\n",
    "    df = df.set_index('transcript_id')\n",
    "    df = df[col_list]\n",
    "    df.to_csv(fn)\n",
    "\n",
    "def create_tissue_directory(sample_data, tissues):\n",
    "    sample_dir = {}\n",
    "    for tissue in tissues:\n",
    "        sample_dir[tissue] = sample_data.loc[sample_data['SMTSD'] == tissue]['SAMPID'].values.tolist()\n",
    "    return(sample_dir)\n",
    "\n",
    "\n",
    "\n",
    "def generate_bicodon_table(refseqs,weights,file_id):\n",
    "\n",
    "    bcds_sorted = bcds.loc[refseqs,:]\n",
    "    bcds_weighted = (bcds_sorted.T * weights).sum(axis=1)\n",
    "    bcds_weighted.index = bcds_weighted.index.str.upper()\n",
    "    bcds_weighted.to_frame().transpose().to_csv(\n",
    "        'bicodon_tables/GTEx_{}_{}_v7.tsv'.format(tissue,file_id), sep='\\t', index=False)\n",
    "    print('Generated Bicodon Table')\n",
    "    return(bcds_weighted)\n",
    "\n",
    "\n",
    "def generate_codon_table(refseqs,weights,file_id):\n",
    "\n",
    "    cds_sorted = cds.loc[refseqs,:]\n",
    "    cds_weighted = (cds_sorted.T * weights).sum(axis=1)\n",
    "\n",
    "    # output codon freq table\n",
    "    output = pd.read_csv('/grid/home/nbourgeois/codonOpt/references/codon_usage.getex.txt',sep='\\t')\n",
    "    output = output.drop(columns='Fraction')\n",
    "    output.loc[:,'Number'] = cds_weighted.values.astype(int)\n",
    "    output.loc[:,'Frequency'] = round(output['Number'] / sum(output['Number']) * 1000,1)\n",
    "    output.to_csv('codon_tables/GTEx_{}_{}_v7.tsv'.format(tissue,file_id), sep='\\t', index=False)\n",
    "    print('Generated Codon Table')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "ref_dir = os.path.join(os.getcwd(),'..','references')\n",
    "os.chdir('/grid/home/nbourgeois/codonOpt/tissue_cocoputs/data')\n",
    "\n",
    "# Original Data\n",
    "gtex_data = pd.read_csv('TPM/GTEx_Lung_TPM_orig.tsv',sep='\\t')\n",
    "gtex_samp_data= pd.read_csv('ref/GTEx_v7_Annotations_SampleAttributesDS.txt',sep='\\t')\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "gtex_samp_data = gtex_samp_data.loc[gtex_samp_data['SAMPID'].isin(gtex_data.columns)]\n",
    "\n",
    "tissues = gtex_samp_data['SMTSD'].unique().tolist()\n",
    "sample_dir = create_tissue_directory(gtex_samp_data,tissues)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "bcds = pd.read_csv('o586358-Human_CDS_Bicod.tsv',sep='\\t')\n",
    "bcds = bcds.drop(columns = ['Protein ID', 'Accession', 'Division', 'Assembly', 'Taxid',\n",
    "                        'Species', 'Organelle', '# Codon Pairs', 'GC%', 'GC1%', 'GC2%', 'GC3%','Unnamed: 4109'])\n",
    "bcds = bcds.groupby('Gene ID').sum()\n",
    "\n",
    "# convert protein info to ensmbl transcript info\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tttttt</th>\n",
       "      <th>tttttc</th>\n",
       "      <th>ttttta</th>\n",
       "      <th>tttttg</th>\n",
       "      <th>tttctt</th>\n",
       "      <th>tttctc</th>\n",
       "      <th>tttcta</th>\n",
       "      <th>tttctg</th>\n",
       "      <th>tttatt</th>\n",
       "      <th>tttatc</th>\n",
       "      <th>...</th>\n",
       "      <th>gggcga</th>\n",
       "      <th>gggcgg</th>\n",
       "      <th>gggagt</th>\n",
       "      <th>gggagc</th>\n",
       "      <th>gggaga</th>\n",
       "      <th>gggagg</th>\n",
       "      <th>gggggt</th>\n",
       "      <th>gggggc</th>\n",
       "      <th>ggggga</th>\n",
       "      <th>gggggg</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ML1</th>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A3GALT2</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11A</th>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11B</th>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>...</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20239 rows × 4096 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         tttttt  tttttc  ttttta  tttttg  tttctt  tttctc  tttcta  tttctg  \\\n",
       "Gene ID                                                                   \n",
       "A1BG          0       0       0       0       0       1       0       2   \n",
       "A1CF          0       0       0       0       0       0       0       0   \n",
       "A2M           0       0       0       0       0       0       0       0   \n",
       "A2ML1         0       7       0       0       6       0       0       6   \n",
       "A3GALT2       0       0       0       0       0       0       0       1   \n",
       "...         ...     ...     ...     ...     ...     ...     ...     ...   \n",
       "ZYG11A        0       6       0       0       0      10       0       0   \n",
       "ZYG11B        0       2       0       0       0       0       0       0   \n",
       "ZYX           0       0       0       0       0       0       0       0   \n",
       "ZZEF1         0       0       0       6      10       0       0      16   \n",
       "ZZZ3          0       0       0       0       0       0       0       0   \n",
       "\n",
       "         tttatt  tttatc  ...  gggcga  gggcgg  gggagt  gggagc  gggaga  gggagg  \\\n",
       "Gene ID                  ...                                                   \n",
       "A1BG          0       0  ...       0       0       0       0       0       0   \n",
       "A1CF         11       0  ...       0       0       0       0       0       0   \n",
       "A2M           0       0  ...       0       5       0       5       0       0   \n",
       "A2ML1         0       0  ...       0       0       0       7       0       0   \n",
       "A3GALT2       0       0  ...       0       0       0       0       0       0   \n",
       "...         ...     ...  ...     ...     ...     ...     ...     ...     ...   \n",
       "ZYG11A        0       0  ...       0       0       0       0       0       0   \n",
       "ZYG11B        3       0  ...       0       0       0       0       3       2   \n",
       "ZYX           0       0  ...       0       0       0       0       0       0   \n",
       "ZZEF1         0      10  ...       6       5       0       0       0       5   \n",
       "ZZZ3          0       0  ...       0       0       0       0       5       0   \n",
       "\n",
       "         gggggt  gggggc  ggggga  gggggg  \n",
       "Gene ID                                  \n",
       "A1BG          0       1       0       0  \n",
       "A1CF          0      12       0       0  \n",
       "A2M           0       0       0       0  \n",
       "A2ML1         0      14       0       0  \n",
       "A3GALT2       0       2       0       0  \n",
       "...         ...     ...     ...     ...  \n",
       "ZYG11A        0      10       0       0  \n",
       "ZYG11B        0       0       0       0  \n",
       "ZYX           0       7       0       0  \n",
       "ZZEF1         0       5       0       0  \n",
       "ZZZ3          0       0       0       0  \n",
       "\n",
       "[20239 rows x 4096 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bcds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "querying 1-1000...done.\n",
      "querying 1001-2000...done.\n",
      "querying 2001-3000...done.\n",
      "querying 3001-4000...done.\n",
      "querying 4001-5000...done.\n",
      "querying 5001-6000...done.\n",
      "querying 6001-7000...done.\n",
      "querying 7001-8000...done.\n",
      "querying 8001-9000...done.\n",
      "querying 9001-10000...done.\n",
      "querying 10001-11000...done.\n",
      "querying 11001-12000...done.\n",
      "querying 12001-13000...done.\n",
      "querying 13001-14000...done.\n",
      "querying 14001-15000...done.\n",
      "querying 15001-16000...done.\n",
      "querying 16001-17000...done.\n",
      "querying 17001-18000...done.\n",
      "querying 18001-19000...done.\n",
      "querying 19001-20000...done.\n",
      "querying 20001-20239...done.\n",
      "Finished.\n",
      "430 input query terms found dup hits:\n",
      "\t[('ABO', 3), ('AGAP11', 2), ('ALG1L', 2), ('ARMCX5-GPRASP2', 2), ('ATP6AP1L', 2), ('BAGE2', 2), ('BO\n",
      "336 input query terms found no hit:\n",
      "\t['ACPP', 'ADPRHL2', 'ANKRD20A2', 'ANKRD20A3', 'ANKRD20A4', 'ARMC4', 'ATP5MD', 'ATP5MPL', 'BHMG1', 'C\n",
      "Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n"
     ]
    }
   ],
   "source": [
    "mg = mygene.MyGeneInfo()\n",
    "geneSyms = mg.querymany(bcds.index.values, scopes='symbol',\n",
    "                        fields='ensembl.gene', species='human')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "tissue = 'Lung'\n",
    "\n",
    "# subset to specific tissue\n",
    "cols = ['transcript_id','gene_id'] + sample_dir[tissue]\n",
    "gtex_data_tissue = gtex_data[cols]\n",
    "tissue = tissue.replace(\" - \", \"_\").replace(\" \", \"_\").replace(\"_(BA9)\", \"\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/data/software/anaconda3/envs/bfx-codon/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  \n"
     ]
    }
   ],
   "source": [
    "# reformat\n",
    "gtex_data_tissue['gene_id'] = gtex_data_tissue['gene_id'].str.split('.',expand=True)[0]\n",
    "gtex_data_tissue_summed = gtex_data_tissue.groupby('gene_id').sum()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get average per transcript\n",
    "gtex_avg = gtex_data_tissue_summed.mean(axis=1)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_weights(geneSyms, gtex_avg):\n",
    "\n",
    "    def append_results(gene):\n",
    "        if gene in gtex_avg.index:\n",
    "            weight = gtex_avg[gene]\n",
    "            symbols.append(gdict['query'])\n",
    "            weights.append(weight)\n",
    "\n",
    "        \n",
    "    symbols = []\n",
    "    weights = []\n",
    "    counter = 0\n",
    "    for gdict in geneSyms:\n",
    "        if 'ensembl' in gdict:\n",
    "            if len(gdict['ensembl']) == 1:\n",
    "                gene = gdict['ensembl']['gene']\n",
    "                append_results(gene)\n",
    "\n",
    "            else:\n",
    "                for ensembl in gdict['ensembl']:\n",
    "                    gene = ensembl['gene']\n",
    "                    append_results(gene)\n",
    "\n",
    "        counter +=1    \n",
    "\n",
    "    return(symbols,weights)\n",
    "\n",
    "\n",
    "symbols,weights = get_weights(geneSyms, gtex_avg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "symbols = []\n",
    "weights = []\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generated Bicodon Table\n"
     ]
    }
   ],
   "source": [
    "bicod_tab = generate_bicodon_table(symbols,weights,'bicodon_test')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "TTTTTT    0.0\n",
       "TTTTTC    0.0\n",
       "TTTTTA    0.0\n",
       "TTTTTG    0.0\n",
       "TTTCTT    0.0\n",
       "         ... \n",
       "GGGAGG    0.0\n",
       "GGGGGT    0.0\n",
       "GGGGGC    0.0\n",
       "GGGGGA    0.0\n",
       "GGGGGG    0.0\n",
       "Length: 4096, dtype: float64"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bicod_tab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TTTTTT</th>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTTTC</th>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTTTA</th>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTTTG</th>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTCTT</th>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGGAGG</th>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGGGGT</th>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGGGGC</th>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGGGGA</th>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGGGGG</th>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4096 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0\n",
       "TTTTTT  0.0\n",
       "TTTTTC  0.0\n",
       "TTTTTA  0.0\n",
       "TTTTTG  0.0\n",
       "TTTCTT  0.0\n",
       "...     ...\n",
       "GGGAGG  0.0\n",
       "GGGGGT  0.0\n",
       "GGGGGC  0.0\n",
       "GGGGGA  0.0\n",
       "GGGGGG  0.0\n",
       "\n",
       "[4096 rows x 1 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bcds_sorted = bcds.loc[symbols,:]\n",
    "bcds_weighted = (bcds_sorted.T * weights).sum(axis=1)\n",
    "bcds_weighted.index = bcds_weighted.index.str.upper()\n",
    "bcds_weighted.to_frame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "weights[0:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'refseqs' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_143861/3243995756.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mbcds\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mrefseqs\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mT\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msum\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'refseqs' is not defined"
     ]
    }
   ],
   "source": [
    "bcds.loc[refseqs,:].T.sum(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "(bcds_sorted.T * weights).sum(axis=1).sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.read_csv('../../references/CoCoPUTs_codon_usage/bicodon_usage/Liver.bicodon.txt',sep='\\t').T.sum()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (codon)",
   "language": "python",
   "name": "bfx-codon"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
